src/python/WMCore/MicroService/MSPileup/MSPileupTasks.py

```diff
@@ -100,17 +105,131 @@ def pileupSizeTask(self):
+                # we should use it from the custom dataset (customName) - if existent
+                cname = doc.get('customName', '')
+                if cname:
+                    blockNames = self.rucioClient.getBlocksInContainer(cname)
+
+            # get portion of DIDs based on ceil(containerFraction * num_rucio_datasets)
+            portion = blockNames[:math.ceil(fraction * len(blockNames))]
+            self.logger.info("For pileup document %s take portion %s", doc, portion)
+
+            # create new container DID in Rucio for our custom name
+            self.rucioClient.createContainer(doc['customName'], scope=self.customRucioScope)
+
+            # we call attachDIDs Rucio wrapper API with our set of DIDs
+            newRules = []
+            for rse in doc['currentRSEs']:
+                self.rucioClient.attachDIDs(rse, doc['customName'], portion, scope=self.customRucioScope)
+
+                # create new rule for custom DID using pileup document rse
+                newRules += self.rucioClient.createReplicationRules(portion, rse)
+
+            self.logger.info("newRules %s", newRules)
+            # set expiration date (to be 24h) for already existing ruleIds from pileup document
+            for rid in doc['ruleIds']:
+                # set expiration date to be 24h ahead of right now
+                opts = {'lifetime': 24 * 60 * 60}
+                self.rucioClient.updateRule(rid, opts)
+
+            # update pileup document
+            if newRules:
+                # remove duplicate rules
+                newRules = set(newRules)
+                try:
+                    # update rules
+                    doc['ruleIds'] += newRules
+
+                    # update transition record of the pileup document, see logic:
+                    # https://gist.github.com/amaltaro/b4f9bafc0b58c10092a0735c635538b5
+
+                    # we should always have transition record(s)
+                    transition = doc['transition']
+                    # take user DN from last record
+                    userDN = transition[-1]['DN']
+                    prevTranRecord = transition[-1]
+                    # generate new custom name
+                    cname = prevTranRecord['customDID']
+                    transitionRecord = {'containerFraction': fraction,
+                                        'customDID': cname,
+                                        'updateTime': gmtimeSeconds(),
+                                        'DN': userDN}
+                    prevTranRecord.update(transitionRecord)
+                    transition[-1] = prevTranRecord
+                    # add new transition record and update custom name
+                    doc['transition'] = transition
+                    doc['customName'] = cname
+                    self.logger.info("update pileup document %s", doc)
+
+                    # update MSPileup document in MongoDB
+                    self.mgr.updatePileup(doc, rseList=newRules)
```

Comment:
bug? providing rules for rseList param(?)

Reply:
yes, it should be `doc['expectedRSEs']` or `doc['currentRSEs']` since my default rseList is None and passed document may have either expectedRSEs or currentRSEs to be not empty lists and during validation we check for them. Please clarify which list should be used.

Reply:
you probably forgot pushing your changes in.